docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+   Making efficient use of padding bits as discriminators seems very difficult and so is likely to be left to a later ABI, if ever implemented.
+
+Once the payload sizes are all known, the algorithm for implementing the packing depends on the type of the largest payload.
+There are two special cases:
+
+If the largest type is a floating-point type and the payload of all other types plus the bits required for the discriminator will fit in the unused bits of the significand of a signalling NaN minus one then the top bit of the significand is reserved to discriminate between a real signalling NaN and another value and the remainder of the types are encoded in the spare bits, starting this algorithm again.
+If the remaining bits of the significand are not sufficient to encode the remaining states then this rule is ignored.
+
+If the largest type is a CHERI capability then the tag bit is used to differentiate pointer and non-pointer types.
+This leaves 128 bits to use for the remainder of the values.
+
+The size of the final object is the number of payload bits, plus the base-two logarithm of the number of discriminator states rounded up.
+The alignment of the result is the alignment of the most strongly aligned type.
+
+If one of the types is a pointer, the object size is pointer sized, and the discriminator is three or fewer bits, then the low bits of the pointer are used for the discriminator.
+Pointer values are stored with their bits in the natural positions and must have the low bits subtracted away on use.
+Other values are stored in the most significant bits.
+
+*Rationale:* Most ISAs make subtracting a 3-bit immediate trivial (often something that can be folded into another instruction).
+Other values that fit in the payload space can be shifted down as required in a single-cycle instruction.
+
+If the payload is an integer number of bytes, the payload is stored first (independent of byte order) so that the payload can benefit from strong alignment, which avoids the payload accidentally spanning a cache line.
+In this case, all elements of the payload are stored starting from the beginning of the payload.
+
+*Rationale:* The payload can be loaded with a single aligned load on any architecture and depending on the target byte order then needs to do either a single zero-extending shift right or a left shift followed by a zero-extending right shift.
+
+If storing the discriminator bits at either end of the space results in different numbers of complete bytes for the payload then the discriminator should stored at the end that results in the larger number of complete bytes for the payload.
+If there are no discriminator bits, this rule is applied as if there were a single bit.
+Payload values that are smaller than the payload space are then stored aligned to the opposite end of the value to the discriminator.
```

Comment:
I'm not sure what the example is here. This paragraph is confusing. Perhaps adding a simple example for each of these topics with rationales will help.

Reply:
Is there a GitHub syntax highlighting thing that makes it easy to write bit layouts?

Reply:
Tables?